### PR TITLE
Implement S3949 logic

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Part of #4631

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/CalculationsShouldNotOverflowBase.cs [hunk 1]
```diff
@@ -18,12 +18,82 @@
...
+            else if (number.Min < typeMin)
+            {
+                ReportIssue(operation, MessageLikely, MessageUnderflow, typeMax);
+            }
+        }
+
+        return context.State;
+    }
+
+    private static SymbolicValue CanOverflow(ProgramState state, IOperation operation) => operation.Kind switch
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/CalculationsShouldNotOverflowBase.cs:63]: I initially wanted to return operation, but that's not possible due to the tricky post increment/decrement :/
So let's give it a better name, as `CanOverflow` answer a question that should return `bool`.

```suggestion
    private static SymbolicValue OverflowCandidateValue(ProgramState state, IOperation operation) => operation.Kind switch
```
or
```suggestion
    private static SymbolicValue OverflowValue(ProgramState state, IOperation operation) => operation.Kind switch
```
or something like that

Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/CalculationsShouldNotOverflowBase.cs:63]: EOL Formatting
```suggestion
    private static SymbolicValue CanOverflow(ProgramState state, IOperation operation) => 
        operation.Kind switch
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/CalculationsShouldNotOverflowBase.cs [hunk 2]
```diff
@@ -18,12 +18,82 @@
...
+        OperationKindEx.Binary when CanOverflow(operation.ToBinary().OperatorKind) => state[operation],
+        OperationKindEx.CompoundAssignment when CanOverflow(operation.ToCompoundAssignment().OperatorKind) => state[operation],
+        OperationKindEx.Increment or OperationKindEx.Decrement when operation.ToIncrementOrDecrement().Target.TrackedSymbol() is { } symbol => state[symbol],
+        _ => null
+    };
+
+    private static bool CanOverflow(BinaryOperatorKind kind) =>
+        kind is BinaryOperatorKind.Add or BinaryOperatorKind.Subtract or BinaryOperatorKind.Multiply or BinaryOperatorKind.Power;
+
+    private static BigInteger? Min(ITypeSymbol type) => type switch
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/CalculationsShouldNotOverflowBase.cs:74]: EOL formatting
```suggestion
    private static BigInteger? Min(ITypeSymbol type) => 
        type switch
```

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/CalculationsShouldNotOverflowBase.cs:74]: Also below

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/CalculationsShouldNotOverflowBase.cs [hunk 3]
```diff
@@ -18,12 +18,82 @@
...
+    protected override ProgramState PostProcessSimple(SymbolicContext context)
+    {
+        if (context.Operation.Instance is var operation
+            && CanOverflow(context.State, operation)?.Constraint<NumberConstraint>() is { } number
+            && Min(operation.Type) is { } typeMin
+            && Max(operation.Type) is { } typeMax)
+        {
+            if (number.Max < typeMin)
+            {
+                ReportIssue(operation, MessageGuaranteed, MessageUnderflow, typeMax);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/CalculationsShouldNotOverflowBase.cs:44]: All message variants needs to be asserted
```suggestion
                ReportIssue(operation, MessageGuaranteed, MessageUnderflow, typeMin);
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/CalculationsShouldNotOverflowBase.cs [hunk 4]
```diff
@@ -18,12 +18,82 @@
...
+            {
+                ReportIssue(operation, MessageGuaranteed, MessageOverflow, typeMax);
+            }
+            else if (number.Max > typeMax)
+            {
+                ReportIssue(operation, MessageLikely, MessageOverflow, typeMax);
+            }
+            else if (number.Min < typeMin)
+            {
+                ReportIssue(operation, MessageLikely, MessageUnderflow, typeMax);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/CalculationsShouldNotOverflowBase.cs:56]: All message variants needs to be asserted
```suggestion
                ReportIssue(operation, MessageLikely, MessageUnderflow, typeMin);
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/CalculationsShouldNotOverflow.cs [hunk 5]
```diff
@@ -186,7 +234,7 @@ public int Overflow1()
         {
             unchecked
             {
-                int i = 1834567890 + 1834567890;    // FIXME Non-compliant
+                int i = 1834567890 + 1834567890;    // Noncompliant
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/CalculationsShouldNotOverflow.cs:237]:
```suggestion
                int i = 1834567890 + 1834567890;    // Noncompliant FP, we don't want to run for methods with unchecked
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/CalculationsShouldNotOverflow.cs [hunk 6]
```diff
@@ -5,73 +5,117 @@ public class Sample
...
 
         ulong ui64 = ulong.MaxValue;
-        _ = ui64 + 1;           // FIXME Non-compliant
+        ui64++;                 // Noncompliant
         ui64 = ulong.MinValue;
-        _ = ui64 - 1;           // FIXME Non-compliant
+        ui64--;                 // Noncompliant
+    }
+
+    public void Upcast()
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/CalculationsShouldNotOverflow.cs:48]: There should be few UTs with
```
var i = Unknown();
if(i > 2147483600)
  _ = i + 100; // Noncompliant
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/CalculationsShouldNotOverflow.cs [hunk 7]
```diff
@@ -86,9 +130,10 @@ public void AssignmentOperators()
         i = 2147483600;
         i *= 100;               // FIXME Non-compliant
 
+        var j = 10;
         i = 2147483600;
-        i /= 10;
-        _ = i * 100;            // FIXME Non-compliant 
+        i /= j;
+        _ = i * 100;            // Noncompliant FIXME wrong reason, i doesn't change because compoundassignment is ignored
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/CalculationsShouldNotOverflow.cs:136]: #7223 is already in review to fix this. 
Once merged, I'd rebase feature branch. And then you can rebase this PR.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Waaay nicer code than whatever the old implementation had to have :D

File: analyzers/its/expected/Net7/Net7--net7.0-S3949.json [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+"location":  {
+"uri":  "sources\Net7\Net7\features\CheckedAndUncheckedUserDefinedOperators.cs",
+"region":  {
+"startLine":  10,
+"startColumn":  41,
+"endLine":  10,
+"endColumn":  46
+}
+}
+},
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/Net7/Net7--net7.0-S3949.json:15]: FN, unchecked keyword

File: analyzers/its/expected/Net7/Net7--net7.0-S3949.json
Comment (Tim-Pohlmann) [inline analyzers/its/expected/Net7/Net7--net7.0-S3949.json:1]: FNs, unchecked keyword

File: analyzers/its/expected/akka.net/Akka--netstandard2.0-S3949.json
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka--netstandard2.0-S3949.json:1]: FN, GetHashCode/unchecked keyword

File: analyzers/its/expected/Net7/Net7--net7.0-S3949.json
Comment (pavel-mikula-sonarsource) [reply analyzers/its/expected/Net7/Net7--net7.0-S3949.json:1]: FP instead of FN?

File: analyzers/its/expected/akka.net/Akka--netstandard2.0-S3949.json
Comment (pavel-mikula-sonarsource) [reply analyzers/its/expected/akka.net/Akka--netstandard2.0-S3949.json:1]: FP

Comment (Tim-Pohlmann) [reply analyzers/its/expected/akka.net/Akka--netstandard2.0-S3949.json:1]: yes, mb!

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM, pleae rebase on feature branch to fix the compound cases in a new commit

**Pull request merged**